blockchain/reactor_test.go

```diff
@@ -0,0 +1,218 @@
+package blockchain
+
+import (
+	"bytes"
+	"testing"
+
+	cfg "github.com/tendermint/tendermint/config"
+	"github.com/tendermint/tendermint/p2p"
+	sm "github.com/tendermint/tendermint/state"
+	"github.com/tendermint/tendermint/types"
+	cmn "github.com/tendermint/tmlibs/common"
+	"github.com/tendermint/tmlibs/db"
+	"github.com/tendermint/tmlibs/log"
+)
+
+func newBlockchainReactor(logger log.Logger, maxBlockHeight int) *BlockchainReactor {
+	config := cfg.ResetTestRoot("node_node_test")
+
+	blockStoreDB := db.NewDB("blockstore", config.DBBackend, config.DBDir())
+	blockStore := NewBlockStore(blockStoreDB)
+
+	stateLogger := logger.With("module", "state")
+
+	// Get State
+	stateDB := db.NewDB("state", config.DBBackend, config.DBDir())
+	state := sm.GetState(stateDB, config.GenesisFile())
+	state.SetLogger(stateLogger)
+	state.Save()
+
+	// Make the blockchainReactor itself
+	fastSync := true
+	bcReactor := NewBlockchainReactor(state.Copy(), nil, blockStore, fastSync)
+
+	// Next: we need to set a switch in order for peers to be added in
+	bcReactor.Switch = p2p.NewSwitch(cfg.DefaultP2PConfig())
+	bcReactor.SetLogger(logger.With("module", "blockchain"))
+
+	// Lastly: let's add some blocks in
+	for blockHeight := 1; blockHeight <= maxBlockHeight; blockHeight++ {
+		firstBlock := makeBlock(blockHeight, state)
+		secondBlock := makeBlock(blockHeight+1, state)
+		firstParts := firstBlock.MakePartSet(types.DefaultBlockPartSize)
+		blockStore.SaveBlock(firstBlock, firstParts, secondBlock.LastCommit)
+	}
+
+	return bcReactor
+}
+
+func newbcrTestPeer(key string) *bcrTestPeer {
+	return &bcrTestPeer{
+		key: key,
+		kvm: make(map[byte]interface{}),
+		ch:  make(chan *keyValue, 2),
+	}
+}
+
+func TestNoBlockMessageResponse(t *testing.T) {
+	logBuf := new(bytes.Buffer)
+	logger := log.NewTMLogger(logBuf)
+	maxBlockHeight := 20
+
+	bcr := newBlockchainReactor(logger, maxBlockHeight)
+	go bcr.OnStart()
+	defer bcr.Stop()
+
+	// Add some peers in
+	peer := newbcrTestPeer(cmn.RandStr(12))
+	bcr.AddPeer(peer)
+
+	chID := byte(0x01)
+	
+	tests := [...]struct {
+		height   int
+		existent bool
+	}{
+		0: {
+			height:   maxBlockHeight + 2,
+			existent: false,
+		},
+		1: {
+			height:   10,
+			existent: true, // We should have this
+		},
+		2: {
+			height:   1,
+			existent: true, // We should have this
+		},
+		3: {
+			height:   100,
+			existent: false,
+		},
+	}
+
+	// Currently the repsonses take asynchronous paths so
+	// the results are skewed. However, we can ensure that
+	// the (expectedHeight, expectedExistence) all tally up.
+	heightTally := map[int]bool{}
```

Comment:
could you help me understand why we need this tally and the code below?

can't we just test that if we call `Receive(bcBlockRequestMessage{-100})` we will get `bsNoBlockResponseMessage`?

```
(p testPeer) TrySend() {
  map[channel] = message
}

bcr.Receive(bcBlockRequestMessage{-100})
assert.Equal(p.map[blockchainChannel], bcNoBlockResponse)
```

Reply:
For sure, so in the tests we use different heights https://github.com/tendermint/tendermint/pull/648/files#diff-4b45e760192d8a11e9b4e278a8c8ea32R72.
Responses are sent with different heights echoed back, the channels might be the same for the same type of response, but the heights vary thus for different heights we strictly expect a response back. Does that make sense?